weave/workers/weave_scorer.py

```diff
@@ -0,0 +1,504 @@
+
+    monitors = fetch_active_monitors(project_id)
+
+    MonitorsCache.set(project_id, monitors)
+
+    return monitors
+
+
+def _make_active_monitor(project_id: str, obj: tsi.ObjSchema) -> ActiveMonitor:
+    return ActiveMonitor(
+        monitor=Monitor(
+            name=obj.val["name"],
+            description=obj.val["description"],
+            sampling_rate=obj.val["sampling_rate"],
+            scorers=resolve_scorer_refs(obj.val["scorers"], project_id),
+            op_names=obj.val["op_names"],
+            query=obj.val["query"],
+            active=obj.val["active"],
+        ),
+        internal_ref=InternalObjectRef(
+            project_id=project_id, name=obj.val["name"], version=obj.digest
+        ),
+        wb_user_id=obj.wb_user_id,
+    )
+
+
+def fetch_active_monitors(project_id: str) -> list[ActiveMonitor]:
+    """Returns active monitors for a given project."""
+    obj_query = tsi.ObjQueryReq(
+        project_id=project_id,
+        filter=tsi.ObjectVersionFilter(
+            is_op=False,
+            base_object_classes=[Monitor.__name__],
+            latest_only=True,
+        ),
+    )
+
+    server = get_trace_server()
+
+    monitor_objects = server.objs_query(obj_query)
+
+    active_monitors = [
+        _make_active_monitor(project_id, obj)
+        for obj in monitor_objects.objs
+        if obj.val["active"]
+    ]
+
+    return active_monitors
+
+
+def resolve_scorer_refs(scorer_ref_uris: list[str], project_id: str) -> list[Scorer]:
+    """Resolves scorer references to Scorer objects."""
+    server = get_trace_server()
+
+    scorer_refs = [
+        parse_internal_uri(scorer_ref_uri) for scorer_ref_uri in scorer_ref_uris
+    ]
+
+    scorer_dicts = server.refs_read_batch(
+        tsi.RefsReadBatchReq(refs=scorer_ref_uris)
+    ).vals
+
+    scorers = [
+        maybe_objectify(from_json(scorer_dict, project_id, server))
+        for scorer_dict in scorer_dicts
+    ]
+
+    for scorer, scorer_ref in zip(scorers, scorer_refs):
+        scorer.__dict__["internal_ref"] = scorer_ref
+
+    return scorers  # type: ignore
+
+
+class CallNotWrittenError(Exception):
+    pass
+
+
+# We do the following retry logic because ClickHouse does not guarantee that the call will be available immediately.
+@retry(
+    retry=retry_if_exception_type(CallNotWrittenError),
+    wait=wait_fixed(1),
+    stop=stop_after_attempt(10),
+    before=before_log(logger, logging.INFO),
+)
+async def get_filtered_calls(
+    project_id: str,
+    call_ids: list[str],
+    op_names: list[str],
+    query: Optional[tsi.Query],
+) -> list[Call]:
+    """This function is used to get the calls that match the filter."""
+    logger.info("Attempting to get %s calls", len(call_ids))
+    server = get_trace_server()
+    count_req = tsi.CallsQueryStatsReq(
+        project_id=project_id,
+        filter=tsi.CallsFilter(call_ids=call_ids),
+        # We want to make sure all end_call events have been written to the DB
+        query={
```

Comment:
since `call_end` events are added to kafka queue after they are adde to CH, i think this filter is going to have very little effect. The comment I made before was that the corresponding `call_start` event may be logged from the client sometime after `call_end`. In those cases, you are going to hit the `raise CallNotWrittenError("Some calls not yet written to DB")` error below - which (while an edge case) can still happen. I think the `CallNotWrittenError` error is pretty harsh. It is going to crash the entire process & drop all the calls currently being processed in the current batch

Reply:
Private convo pointed out that this will retry for a bit, then drop that batch, but not terminate the process